src/ripple/core/JobCounter.h

```diff
@@ -0,0 +1,188 @@
+    // in flight.  This allows Stoppables to hold off declaring stopped()
+    // until all their JobQueue Jobs are dispatched.
+    template <typename JobHandler>
+    class CountedJob
+    {
+    private:
+        JobCounter& jobCounter_;
+        JobHandler jobHandler_;
+
+        // A helper for the following static_assert.
+        template <typename T>
+        static typename std::remove_reference<T>::type&
+        toLvalueRef (T&& arg)
+        {
+            using ReturnType = typename std::remove_reference<T>::type&;
+            return static_cast<ReturnType>(arg);
+        }
+
+        static_assert (
+            std::is_same<decltype(
+                jobHandler_(toLvalueRef(Job{}))), void>::value,
+                "JobHandler must be callable with Job&");
+
+    public:
+        CountedJob () = delete;
+
+        CountedJob (CountedJob const& rhs)
+        : jobCounter_ (rhs.jobCounter_)
+        , jobHandler_ (rhs.jobHandler_)
+        {
+            jobCounter_.adjustCount (1);
+        }
+
+        CountedJob (CountedJob&& rhs)
+        : jobCounter_ (rhs.jobCounter_)
+        , jobHandler_ (std::move (rhs.jobHandler_))
+        {
+            jobCounter_.adjustCount (1);
+        }
+
+        CountedJob (JobCounter& jobCounter, JobHandler&& jobHandler)
+        : jobCounter_ (jobCounter)
+        , jobHandler_ (std::move (jobHandler))
+        {
+            jobCounter_.adjustCount (1);
+        }
+
+        CountedJob& operator=(CountedJob const& rhs) = delete;
+        CountedJob& operator=(CountedJob&& rhs) = delete;
+
+        ~CountedJob ()
+        {
+            jobCounter_.adjustCount (-1);
+        }
+
+        void operator ()(Job& job)
+        {
+            return jobHandler_ (job);
+        }
+    };
+
+public:
+    JobCounter() = default;
+    // Not copyable because outstanding counts would be hard to sort out.
+    JobCounter (JobCounter const&) = delete;
+    // Not movable because in-flight counted Jobs have a pointer to this.
+    JobCounter (JobCounter&&) = delete;
+
+    JobCounter& operator=(JobCounter const&) = delete;
+
+    /** Destructor verifies all in-flight jobs are complete. */
+    ~JobCounter()
+    {
+        join();
+    }
+
+    /** Returns once all counted in-flight Jobs are destroyed. */
+    void join()
+    {
+        std::unique_lock<std::mutex> lock {mutex_};
+        waitForJobs_ = true;
+        if (jobCount_ > 0)
+        {
+            allJobsDoneCond_.wait (lock, [this] { return jobCount_ == 0; });
+        }
+    }
+
+    /** Wrap the passed lambda with a reference counter.
+
+        @param jobHandler Lambda that accepts a Job& parameter and returns void.
+        @return If join() has been called returns boost::none.  Otherwise
+                returns a boost::optional that wraps jobHandler with a
+                reference counter.
+    */
+    template <class JobHandler>
+    boost::optional<CountedJob<JobHandler>>
+    wrap (JobHandler&& jobHandler)
```

Comment:
I'm changing the variable names to `handler` and `counter`.  That should help everyone with tired eyes.